amaranth/vendor/gowin.py

```diff
@@ -0,0 +1,548 @@
+    def _oscillator(self):
+        oscillators = []
+        for osc,devs in self.osc_dev_mapping.items():
+            if self.device in devs:
+                oscillators.append(osc)
+        # should this be unique?
+        if len(oscillators) != 1:
+            assert False
+        return oscillators[0]
+
+
+    ### Common templates
+
+    _common_file_templates = {
+        "{{name}}.cst": r"""
+            // {{autogenerated}}
+            {% for port_name, pin_name, attrs in platform.iter_port_constraints_bits() -%}
+                IO_LOC "{{port_name}}" {{pin_name}};
+                {% for attr_name, attr_value in attrs.items() -%}
+                    IO_PORT "{{port_name}}" {{attr_name}}={{attr_value}};
+                {% endfor %}
+            {% endfor %}
+        """,
+    }
+
+
+    ### Apicula templates
+
+    _apicula_required_tools = ["yosys", "nextpnr-gowin", "gowin_pack"]
+
+    _apicula_file_templates = {
+        **TemplatedPlatform.build_script_templates,
+        **_common_file_templates,
+        "{{name}}.il": r"""
+            # {{autogenerated}}
+            {{emit_rtlil()}}
+        """,
+        "{{name}}.debug.v": r"""
+            /* {{autogenerated}} */
+            {{emit_debug_verilog()}}
+        """,
+        "{{name}}.ys": r"""
+            # {{autogenerated}}
+            {% for file in platform.iter_files(".v") -%}
+                read_verilog {{get_override("read_verilog_opts")|options}} {{file}}
+            {% endfor %}
+            {% for file in platform.iter_files(".sv") -%}
+                read_verilog -sv {{get_override("read_verilog_opts")|options}} {{file}}
+            {% endfor %}
+            {% for file in platform.iter_files(".il") -%}
+                read_ilang {{file}}
+            {% endfor %}
+            read_ilang {{name}}.il
+            delete w:$verilog_initial_trigger
+            {{get_override("script_after_read")|default("# (script_after_read placeholder)")}}
+            synth_gowin {{get_override("synth_opts")|options}} -top {{name}} -json {{name}}.json
+            {{get_override("script_after_synth")|default("# (script_after_synth placeholder)")}}
+        """,
+    }
+
+    _apicula_command_templates = [
+        r"""
+        {{invoke_tool("yosys")}}
+            {{quiet("-q")}}
+            {{get_override("yosys_opts")|options}}
+            -l {{name}}.rpt
+            {{name}}.ys
+        """,
+        r"""
+        {{invoke_tool("nextpnr-gowin")}}
+            {{quiet("--quiet")}}
+            {{get_override("nextpnr_opts")|options}}
+            --log {{name}}.tim
+            --device {{platform._part}}
+            --family {{platform.device}}
+            --json {{name}}.json
+            --cst {{name}}.cst
+            --write {{name}}.json
+        """,
+        r"""
+        {{invoke_tool("gowin_pack")}}
+            -d {{platform.device}}
+            -o {{name}}.fs
+            {{get_override("gowin_pack_opts")|options}}
+            {{name}}.json
```

Comment:
I see that gowin_pack has a `--cst` option--do you know what's that about?

Reply:
As far as I remember, "cst" in GOWIN is for the constraint file, like the "pcf" (port configuration file?) files of other platforms.

From source: https://github.com/YosysHQ/apicula/blob/master/apycula/gowin_pack.py#L997
For instance: https://github.com/YosysHQ/apicula/blob/033d1c6bc8cc93d5a24f9ba935fb690b023149a2/examples/honeycomb.cst#L4

Reply:
Yes, I'm aware. I'm asking what function it has in gowin_pack, considering that it is already used for nextpnr-gowin.

Reply:
My bad for stating the obvious.

It is generated out rather than read in... maybe just for debug?

> Create .cst (constraints) file from gowin_pack and gowin_unpack
> Add the .cst file generation during packing
> Make the .cst building optional
> -- https://github.com/YosysHQ/apicula/commit/a327700441dc367d895e98c0b4d2a7c6fae35024

Nothing about `cst` [before it](https://github.com/YosysHQ/apicula/blob/f0162d0d46d153a5436e904cf50d4be6b3ddc97e/apycula/gowin_pack.py), which suggests it is not required to be there, but convenient to have?

Reply:
Oh, it _generates_ the constraint file. I'm at a loss of why, but this answers it. Thank you!